README.md
# CausalMBSTS

## What is CausalMBSTS?
CausalMBSTS is an R package that can be used to infer a causal effect of an intervention 
on a multivariate time series. For example, if two items are frequently bought together (e.g., printer and ink) but only one of them gets a discount, what is the effect of the discount on the sales of both products? To answer this question, the dependence structure between the items must be considered. The package does so by using Multivariate Bayesian Structural Time Series models (MBSTS). 

## Why MBSTS models?
MBSTS models are flexible tools that also allow for a transparent way to deal with uncertainty. Flexibility comes from an option for a user to add sub-components (e.g., trend, seasonality, and cycle) that encapsulate the characteristics of a time series. Transparency comes from the Bayesian framework, in which our uncertainty surrounding model parameters is explicitly accounted for by setting prior distributions that represent our prior knowledge of the phenomenon. Then, by combining the priors with the data, we obtain posterior distributions that can be used for several purposes (e.g., compute a point estimate, credible intervals, perform prediction). The package implements the Gaussian linear model that, possibly after transformation of the dependent variable, provides an adequate representation for many time series that are analyzed in practice.     

## How does the package work?
The main function included in the package is <tt>CausalMBSTS()</tt>. The user is asked to provide the data, a vector of dates (including the intervention date) and to select the desired model components. The user can set custom values of the hyperparameters of the priors or use the default values. Then, the function uses a Gibbs sampler to draw from the joint posterior distribution of the model parameters before the intervention. The next step is the prediction of the counterfactual outcomes in the absence of intervention, which is done by sampling from the posterior predictive distribution (PPD). The causal effect at each point in time after the intervention is computed as the difference between the observed outcome and the PPD. Plotting and printing methods for the resulting object are provided. 

Aside from causal inference, the package can also be used for forecasting. This task can be completed with the function <tt>as.mbsts()</tt> coupled with the method <tt>predict.mbsts()</tt>: the former estimates the MBSTS model and outputs an object of class <tt>mbsts</tt>; the latter performs a prediction step for a given number of future periods. 

## Installing the package from GitHub

```{r}
# Get CausalMBSTS from GitHub
devtools::install_github("FMenchetti/CausalMBSTS", build_vignettes = TRUE)

# Get CausalMBSTS from CRAN
install.packages("CausalMBSTS")
library(CausalMBSTS)
```

## Further readings
Menchetti \& Bojinov (2020), Estimating causal effects in the presence of partial interference using multivariate Bayesian structural time series models <https://arxiv.org/abs/2006.12269>


NEWS.md
# CausalMBSTS 0.1.1

* Minor fixes, see issue #46 (fixing existing problems from CRAN check results)

# CausalMBSTS 0.1.0

* First release


tests/testthat.R
library(testthat)
library(CausalMBSTS)

test_check("CausalMBSTS")


tests/testthat/test-plot.R
load("./refmod.Rdata")

test_that("plotting works", {
    x <- sapply( c("impact", "forecast", "ppchecks"),
                function(x) try(plot(causal.1, int.date, type = x, prob = .5)))
    if(interactive()) dev.off()
    expect_true(all(class(x) != 'try-error'))
})


tests/testthat/test-predict.R
load("./refmod.Rdata")

test_that("prediction works", {
    mbsts_pred <- try(predict(mbsts.1, steps.ahead = 10))
    expect_true(class(mbsts_pred) != 'try-error')
})


R/as_mbsts.R
#' Definition and estimation of a Multivariate Bayesian Structural Time Series model (MBSTS)
#'
#' The function creates a multivariate Bayesian structural time series model. It then estimates
#' the model, samples from the joint posterior distribution of its parameters, and outputs an
#' object of class \code{mbsts}.
#'
#' @param y t x d data.frame (or matrix) of observations, where d is the number of time series
#' in the multivariate model.
#' @param components Character vector specifying the components of the multivariate structural
#' time series model. Possible values are c("trend", "slope", "seasonal", "cycle").
#' @param seas.period Length of the seasonal pattern, if present.
#' @param cycle.period Length of the cycle pattern, if present.
#' @param X Optional t x N data frame (or matrix) of N predictors.
#' @param H P x P variance-covariance matrix of the regression coefficients. Set by
#' default to H = c(X'X)^(-1) which is akin to the Zellner's g-prior. The value of
#' the scaling factor is set to \code{c = 1}. Alternative priors could be
#' H = c*diag((X'X)^(-1)) or H = c*I. See also Smith & Kohn, 1995 that suggest
#' setting \code{c} in the range [10,1000].
#' @param nu0.r Degrees of freedom of the Inverse-Wishart prior for each element of
#' Sigma.r, a vector of errors for state r. Set by default to d + 2
#' (must be greater than d - 1).
#' @param s0.r Scale matrix of the Inverse-Wishart prior for each Sigma.r, a vector
#' of errors for state r. Must be a (d x d) positive definite. Default set to the
#' variance-covariance matrix of y multiplied by a scaling factor of 0.01.
#' @param nu0.eps Degrees of freedom of the Inverse-Wishart prior for Sigma.eps,
#' a vector of observation errors for each time series. Set by default to d + 2
#' (must be greater than d - 1).
#' @param s0.eps Scale matrix of the Inverse-Wishart prior for Sigma.eps, a vector
#' of observation errors for each time series. Must be a (d x d) positive definite.
#' Default set to Default set to the variance-covariance matrix of y multiplied by
#' a scaling factor of 0.01.
#' @param niter Number of MCMC iterations.
#' @param burn Desired burn-in, set by default to 0.1 * \code{niter}.
#' @param ping A status message is printed every \code{ping} iteration. Default
#'   set to 0.1 * \code{niter}. Set to 0 to not track the status.
#'
#' @return An object of class 'mbsts' which is a list with the following components:
#' \describe{
#'   \item{eta.samples}{(\code{niter}- \code{burn}) draws from the distribution of eta_r.}
#'   \item{eps.samples}{(\code{niter}- \code{burn}) draws from the distribution of eps.}
#'   \item{states.samples}{(\code{niter}- \code{burn}) draws from p(alpha_t | Y_{1:T}).}
#'   \item{Sigma.r}{(\code{niter}- \code{burn}) draws from the posterior distribution of Sigma.r.}
#'   \item{Sigma.eps}{(\code{niter}- \code{burn}) draws from the posterior distribution of Sigma.eps.}
#'   \item{Z.beta}{(\code{niter}- \code{burn}) x P matrix of the models selected at each
#'   iteration (if a matrix of predictors is provided).}
#'   \item{beta}{ P x d x (\code{niter}- \code{burn}) ) array of the draws from the posterior
#'   distribution of the regression coefficient matrix (if a matrix of predictors is provided).}
#'   \item{X}{Predictor matrix (if provided).}
#'   \item{y}{Matrix of observations.}
#'   \item{Z}{(d x m) selection matrix of the observation equation.}
#'   \item{Tt}{(m x m) matrix of the state equation.}
#'   \item{R}{(m x r) matrix selecting the state disturbances.}
#'   \item{niter}{Number of mcmc iterations.}
#'   \item{burn}{Burn-in.}
#'   }
#' @export
#'
#' @examples
#' ## Example 1 : local level + seasonal (d = 3)
#' y <- cbind(seq(0.5,100,by=0.5)*0.1 + rnorm(200),
#'            seq(100.25,150,by=0.25)*0.05 + rnorm(200),
#'            rnorm(200, 5,1))
#' mbsts.1 <- as.mbsts(y = y, components = c("trend", "seasonal"), seas.period = 7,
#'                     s0.r = diag(3), s0.eps = diag(3), niter = 50, burn = 5)
#'
#' ## Example 2 : local level + seasonal + covariates (d = 2)
#' y <- cbind(rnorm(100), rnorm(100, 2, 3))
#' X <- cbind(rnorm(100, 0.5, 1) + 5, rnorm(100, 0.2, 2) - 2)
#' mbsts.2 <- as.mbsts(y = y, components = c("trend", "seasonal"), , seas.period = 7,
#'                     X = X, s0.r = diag(2), s0.eps = diag(2), niter = 100, burn = 10)

as.mbsts <- function(y, components, seas.period = NULL, cycle.period = NULL, X = NULL,
                     H = NULL, nu0.r = NULL, s0.r = 0.01 * var(y, na.rm = TRUE), nu0.eps = NULL,
                     s0.eps = 0.01 * var(y, na.rm = TRUE), niter, burn, ping = NULL){

  ## Parameter checks
  if(!is.matrix(y) && !is.data.frame(y)) stop("`y` must be a matrix or data.frame")
  if(all(!components %in% c("trend", "slope", "seasonal", "cycle")))
    stop("`components` must be one of 'trend', 'slop', 'seasonal', or 'cycle'")
  if(!missing(seas.period) && (!is.numeric(seas.period) ||length(seas.period) != 1))
    stop("`seas.period` must be a numeric vector of length one")
  if(!missing(cycle.period) && (!is.numeric(cycle.period) ||length(cycle.period) != 1))
    stop("`cycle.period` must be a numeric vector of length one")
  if(!missing(X)) {
    if(!is.matrix(X) && !is.data.frame(X)) stop("`X` must be a matrix or data.frame")
    if(nrow(X) != nrow (y)) stop("nrow(X) != nrow(X)")
  }
  if(!missing(s0.r) && (!is.matrix(s0.r) || !all(dim(s0.r) == ncol(y))))
    stop("`s0.r` must be a d x d matrix")
  if(!missing(nu0.eps)  && (!is.numeric(nu0.eps) || length(nu0.eps) != 1 || nu0.eps <= nrow(y)))
    stop("`nu0.eps` must be a length-one numeric vector value >= to d")
  if(!missing(s0.eps) && (!is.matrix(s0.eps) || !all(dim(s0.r) == ncol(y))))
    stop("`s0.eps` must be a d x d matrix")
  if(!missing(niter) && !is.numeric(niter)) stop("`niter` must be a length-one numeric vector")
  if(!missing(ping) && !is.numeric(ping)) stop("`ping` must be a length-one numeric vector")
  # Model definition
  Smodel <- model(y = y, components = components, seas.period = seas.period, cycle.period = cycle.period)

  # Model estimation (MCMC)
  est <- mcmc(Smodel = Smodel, X = X, H = H, nu0.r = nu0.r, s0.r = s0.r, nu0.eps = nu0.eps,
              s0.eps = s0.eps, niter = niter, burn = burn, ping = ping)

  class(est) <- "mbsts"
  return(est)
}


R/model_mbsts.R
#' Multivariate structural time series model definition
#'
#' @param y t x d data.frame (or matrix) of observations, where d is the number of time series
#' in the multivariate model.
#' @param components Character vector specifying the components of the
#'   multivariate structural time series model. Possible values are c("trend",
#'   "slope", "seasonal", "cycle").
#' @param seas.period Length of the seasonal pattern, if present.
#' @param cycle.period Length of the cycle pattern, if present.
#'
#' @return An object of class 'SSModel'.
#' @export
#' @keywords internal
#'
#' @examples
#' # Example 1 : local level + seasonal
#' y <- cbind(seq(0.5,200,by=0.5)*0.1 + rnorm(400),
#'            seq(100.25,200,by=0.25)*0.05 + rnorm(400),
#'            rnorm(400, 5,1))
#' model.1 <- model(y, components = c("trend","seasonal"), seas.period = 7)
#'
#' # Example 2: local level  + cycle
#' t <- seq(from = 0,to = 4*pi, length.out=300)
#' y <- cbind(3*sin(2*t)+rnorm(300), 2*cos(2*t) + rnorm(300))
#' model.2 <- model(y, components = c("trend", "cycle"), cycle.period = 75)

model <- function(y, components, seas.period = NULL, cycle.period = NULL){

  y <- as.matrix(y)

  if("trend" %in% components & "slope" %in% components){

    mt<-paste("SSM","trend","(", "degree = 2", " , ", "Q = list(matrix(NA), matrix(NA))",")", sep="")
  } else if ("trend" %in% components & !("slope" %in% components)) {
    mt<-paste("SSM","trend","(", "degree = 1", " , ", "Q = matrix(NA)",")", sep="")
  } else { mt <- NULL }

  if("seasonal" %in% components){
    if(is.null(seas.period)){ stop("A seasonal model needs a seas.period") }
    ms <- paste("SSM","seasonal", "(", "period = ", seas.period, " , ", "Q = matrix(NA)", ")", sep="")
  } else { ms <- NULL }

  if("cycle" %in% components){
    if(is.null(cycle.period)){ stop("A cyclical model needs a cycle.period") }
    mc <- paste("SSM","cycle", "(", "period = ", cycle.period, " , ", "Q = matrix(NA)", ")", sep="")
  } else { mc <- NULL }

  formula <- as.formula(paste("y ~ ", paste(unlist(list(mt,ms,mc)), collapse = " + ")))

  m <- SSModel(formula)

  return(m)

}


R/predict_mbsts.R
######################################################################################
######################################################################################
####  Author:           Fiammetta Menchetti                                       ####
####                                                                              ####
####  Date last update: 2020-06-23                                                ####
####                                                                              ####
####  Content:          Prediction of a given MBSTS model                         ####
####                                                                              ####
####                                                                              ####
####  Main function :   predict.mbsts                                             ####
####                                                                              ####
######################################################################################
######################################################################################

#' Predictions for a given multivariate Bayesian structural time series model
#'
#' Given an object of class 'mbsts' and the number of 'steps.ahead' in the future to be
#' forecasted, this function provides in-sample forecasts and out-of-sample forecasts,
#' both based on drawing from the posterior predictive distribution. If 'mbsts' contains a
#' regression component, then the new matrix of predictors 'newdata' must be provided.
#' Note that NA values are not allowed in the new regressor matrix.
#'
#' @importFrom MASS mvrnorm
#' @param object An object of class 'mbsts', a result of call to \code{\link{as.mbsts}}.
#' @param steps.ahead An integer value (S) specifying the number of steps ahead
#'     to be forecasted. If 'object' contains a regression component, the argument
#'     is disregarded and a prediction is made with the same length of
#'     'newdata'.
#' @param newdata Optional matrix of new data. Only required when 'object'
#'     contains a regression component.
#' @param ... Arguments passed to other methods (currently unused).
#'
#' @return Returns a list with the following components
#' \describe{
#'   \item{post.pred.0}{t x d x ('niter'- 'burn') array of in-sample forecasts.}
#'   \item{post.pred.1}{S x d x ('niter'- 'burn') array out-of-sample forecasts, where S is
#'         the number of forecasted periods (set to the length of 'steps.ahead' or 'newdata').}
#'   \item{post.pred}{(t + S) x d x ('niter'- 'burn') array combining in- and out-of-sample forecasts.}
#' }
#' @export
#'
#' @examples
#' ## Note: The following are toy examples, for a proper analysis we recommend to run
#' ##       at least 1000 iterations and check the convergence of the Markov chain
#'
#' ## Example 1 :
#' y <- cbind(seq(0.5,100,by=0.5)*0.1 + rnorm(200),
#'            seq(100.25,150,by=0.25)*0.05 + rnorm(200),
#'            rnorm(200, 5,1))
#' mbsts.1 <- as.mbsts(y = y, components = c("trend", "seasonal"), seas.period = 7, s0.r = diag(3),
#'                     s0.eps = diag(3), niter = 50, burn = 5)
#' pred.1 <- predict(mbsts.1, steps.ahead = 10)
#'
#' ## Example 2
#' y <- cbind(rnorm(100), rnorm(100, 2, 3))
#' X <- cbind(rnorm(100, 0.5, 1) + 5, rnorm(100, 0.2, 2) - 2)
#' mbsts.2 <- as.mbsts(y = y, components = c("trend", "seasonal"),
#'                     seas.period = 7, X = X, s0.r = diag(2),
#'                     s0.eps = diag(2), niter = 100, burn = 10)
#' newdata <- cbind(rnorm(30), rt(30, 2))
#' pred.2 <- predict(mbsts.2, newdata = newdata)

predict.mbsts <- function(object, steps.ahead, newdata = NULL, ...) {

    # Given an object of class 'mbsts' and the number of 'steps.ahead' in the future to be
    # forecaste, this function provides in-sample forecasts and out-of-sample forecasts,
    # both based on drawing from the posterior predictive distribution. If 'x' contains
    # a regression component, then the new matrix of predictors 'newdata' must be provided.
    # Note that NA values are not allowed in the new regressor matrix.
    #
    # Args:
    #   object       : an object of class 'mbsts'
    #   steps.ahead : an integer value specifying the number of steps ahead to be forecasted
    #   newdata     : optional matrix of new data
    #
    # Value:
    #   post.pred.0 : T x d x niter array of in-sample forecasts
    #   post.pred.1 : S x d x niter array out-of-sample forecasts, where S is the number of
    #                 forecasted periods (set to the length of provided new data)
    #   post.pred   : (T + S) x d x niter array combining in- and out-of-sample forecasts

    ## Parameter checks
    if(!missing(steps.ahead))
        if(as.integer(steps.ahead) != steps.ahead | length(steps.ahead) != 1)
            stop("'steps.ahead' must be an integer of length one")
    if(!missing(newdata))
        if(!is.null(newdata) && !any(class(newdata) %in% c("matrix", "data.frame")))
            stop("'newdata' must be a matrix")

    ### Dimensionalities & other objects
    mbsts <- object
    # get dim
    t <- dim(mbsts$y)[1]  # number of time points
    rr <- dim(mbsts$R)[2]  # tot number of state disurbances
    d <- dim(mbsts$y)[2]  # number of time series
    #r <- rr/d  # number of state disturbances for each time series
    M <- dim(mbsts$Tt)[1]  # tot number of states

    if (is.null(mbsts$X)) {
        step <- steps.ahead
    } else {
        newdata <- as.matrix(newdata)
        step <- dim(newdata)[1]
    }  # number of obs to forecast

    niter <- mbsts$niter - mbsts$burn  # number of simulations
    last <- dim(mbsts$eta.samples)[1]  # last draw from the posterior p(alpha_t|Y_n)


    ### Empty arrays to store iterations
    eta.new <- array(NA, c(step, rr, niter))

    states.new <- array(NA, c(step, M, niter))
    colnames(states.new) <- colnames(mbsts$states.samples)

    y.star <- array(NA, c(step, d, niter))

    post.pred.0 <- array(NA, c(t, d, niter))  # storing in-sample draws
    post.pred.1 <- array(NA, c(step, d, niter))  # storing out-of-samples draws
    post.pred <- array(NA, c(t + step, d, niter))  # all together


    for (i in 1:niter) {

        ### STEP 1: Get in-samples draws from the ppd.

        # Details:  let theta = (alpha, beta, Sigma.eps, z, Sigma.r),
        #           during the MCMC we got samples from the joint posterior distribution p(theta|y). So
        #           now we can draw new values y.tilde from the posterior predictive distribution by simply
        #           taking the 'niter' draws from the joint p(theta|y) and substitute them into model equations
        #           (we assume that y.new comes from the same distribution of y so that y.new is
        #           independent of y given theta).

        if (is.null(mbsts$X)) {
            post.pred.0[, , i] <- tcrossprod(mbsts$states.samples[, , i], mbsts$Z[, , 1]) + mbsts$eps.samples[,
                , i]
        } else {
            post.pred.0[, , i] <- tcrossprod(mbsts$states.samples[, , i], mbsts$Z[, , 1]) + mbsts$X %*%
                mbsts$beta[, , i] + mbsts$eps.samples[, , i]
        }

        ### STEP 2: Get new out-of-samples draws from the ppd.

        # Details: this time an out-of sample draw y.new is no more independent
        #          of past y given theta. To see that, let's say we want to sample y_t+k | Y_t.
        #          This time, theta = (alpha_t+k,...,alpha_t+1, alpha_t, beta, Sigma.eps, z, Sigma.r)
        #          and from our MCMC we have just the posterior of (alpha_t, beta, Sigma.eps, z, Sigma.r)
        #          because we sampled from the full conditional p(alpha_t | Y_t, beta, Sigma.eps, z, Sigma.r).
        #          Thus, to sample from y_t+k | y_t we should 'recover' the missing dependence structure
        #          p(alpha_t+k,...,alpha_t+1 | theta') where theta' is our 'old' theta, that is
        #          theta' = alpha_t, beta, Sigma.eps, z, Sigma.r. More details in the paper.
        #          Conversely, y*_t+k =  y_t+k - Z alpha_t+k is independent of y*_t given theta',
        #          because there's no more time component and we can just write y*.
        #          During the MCMC we sampled from the joint posterior p(beta,Sigma.eps,z|y*),
        #          and as in step 1 we can use those draws to sample from the
        #          posterior predictive distribution y.new* | y* by simply taking the draws from
        #          p(beta,Sigma.eps,z|y*) and then draw from p(y.new*|beta,Sigma.eps,z).
        #

        # 2.1. Sampling new states

        eta.new <- matrix(mvrnorm(1, rep(0, rr), mbsts$Sigma.r[, , i]), nrow = rr)
        states.new[1, , i] <- mbsts$Tt[, , 1] %*% matrix(mbsts$states.samples[last, , i], M) + mbsts$R[,
            , 1] %*% eta.new

        ind <- if (step > 1) {
            2:step
        } else {
            NULL
        }
        for (j in ind) {
            eta.new <- matrix(mvrnorm(1, rep(0, rr), mbsts$Sigma.r[, , i]), nrow = rr)
            states.new[j, , i] <- mbsts$Tt[, , 1] %*% matrix(states.new[j - 1, , i], nrow = M) + mbsts$R[,
                , 1] %*% eta.new
        }

        # 2.2. Sampling y*.new (note that in the simple case without covariates y*.new = eps)

        if (is.null(mbsts$X)) {
            y.star[, , i] <- mvrnorm(1, rep(0, d), mbsts$Sigma.eps[, , i])
        } else {
            y.star[, , i] <- newdata %*% mbsts$beta[, , i] + mvrnorm(1, rep(0, d), mbsts$Sigma.eps[,
                , i])
        }

        # 2.3. Out-of-sample draws

        post.pred.1[, , i] <- states.new[, , i] %*% t(mbsts$Z[, , 1]) + y.star[, , i]

        ### STEP 3: Combining in-sample and out-of-samples draws

        post.pred[, , i] <- rbind(post.pred.0[, , i], post.pred.1[, , i])
    }

    return(list(post.pred.0 = post.pred.0, post.pred.1 = post.pred.1, post.pred = post.pred))
}


R/plot_mbsts.R
#' Plotting function for object of class CausalMBSTS
#'
#' Given an object of class 'CausalMBSTS', the function draws: i) the plot of
#' the estimated (pointwise) causal effect; ii) the original time series plotted
#' against the predicted counterfactual; iii) posterior predictive checks; iv)
#' regressor inclusion probabilities (only for models with a regression
#' component).
#'
#' @importFrom forecast Acf
#' @param x Object of class 'CausalMBSTS'
#' @param int.date Date of the intervention.
#' @param type A character string indicating the type of plot to be produced.
#'   Possible values in 'c('impact', 'forecast', 'ppchecks', 'inclusion.probs')'.
#'   See Details for further explanation.
#' @param prob Regressors inclusion probabilities above 'prob' are plotted.
#'   Optional, only required for type = 'inclusion.prob'.
#' @param ... Arguments passed to other methods (currently unused).
#'
#' @details
#' Option 'impact' for parameter \code{type} plots the general causal effect at every time points in the post
#'   period. Multiple plots will be generated, corresponding to each combination of time series and horizon (if specified).
#'   Option 'forecast' plots the observed time series against the predicted counterfactual, one plot per each
#'   combination of time series and horizon (if specified). 'ppchecks' draws posterior predictive checks for the model
#'   estimated in the pre-period. They include four plots generated for each time series (and horizon). The plots are
#'   (1) density of posterior mean vs. density of the data before intervention, (2) Histogram of maximum in-sample forecasts
#'   and Bayes p-value, (3) QQ-plot of residuals, and (4) ACF of residuals. Option 'inclusion.probs' plots the regressors'
#'   inclusion probabilities above 'prob'.
#'
#' @return NULL, invisibly.
#' @export
#'
#' @examples
#' ## Note: The following are toy examples, for a proper analysis we recommend to run
#' ##       at least 1000 iterations and check the convergence of the Markov chain
#'
#' ## Example 1 (daily data, d = 3, local level + seasonal + covariates)
#' y <- cbind(seq(0.5,100,by=0.5)*0.1 + rnorm(200),
#'            seq(100.25,150,by=0.25)*0.05 + rnorm(200),
#'            seq(1,200,by=1)*(-0.01) + rnorm(200, 0, 0.5))
#' dates <- seq.Date(from = as.Date('2019-01-10'),by = "days", length.out = 200)
#'
#' # Adding a fictional intervention and four covariates. To illustrate the
#' # functioning of Bayesian model selection, one covariate is assumed to be
#' # unrelated to y.
#' int.date <- as.Date('2019-04-01')
#' y.new <- y; y.new[dates >= int.date, ] <- y.new[dates >= int.date, ]*1.3
#' x1 <- y[,1]*0.5 + y[,2]*0.3 + y[,3]*0.1
#' x2 <- y[,2]*0.1 + rnorm(dim(y)[1],0,0.5)
#' x3 <- y[,3]*1.2 + rnorm(dim(y)[1],0,0.5)
#' x4 <- rnorm(dim(y)[1], 5, 10)
#' X <- cbind(x1, x2, x3, x4)
#'
#' # Model definition
#' causal.1 <- CausalMBSTS(y.new, components = c("trend", "seasonal"), seas.period = 7,
#'                         X = X, dates = dates, int.date = int.date,
#'                         s0.r = 0.1*diag(3), s0.eps = 0.1*diag(3), niter = 50,
#'                         burn = 5, horizon = as.Date(c('2019-04-08','2019-07-28')))
#'
#' ## Plotting
#' plot(causal.1, int.date = int.date, type = 'inclusion.probs', prob = 0.1)
#' # as expected, x4 is rarely included in the model
#' oldpar <- par(no.readonly = TRUE)
#' par(mar = c(2,2,2,2))
#' par(mfrow=c(2,3))
#' plot(causal.1, int.date = int.date, type = c('impact', 'forecast'))
#' par(mfrow=c(3,4))
#' plot(causal.1, type = 'ppchecks', int.date = int.date)
#' par(oldpar)
#'
#' ## Example 2
#' set.seed(1)
#' t <- seq(from = 0,to = 4*pi, length.out=300)
#' y <- cbind(3*sin(2*t)+rnorm(300), 2*cos(2*t) + rnorm(300))
#' dates <- seq.Date(from = as.Date("2015-01-01"), by = "week", length.out=300)
#' int.date <- as.Date("2020-02-27")
#' y[dates >= int.date,] <- y[dates >= int.date,]+2
#'
#' # Model definition
#' causal.2 <- CausalMBSTS(y, components = c("trend", "cycle"), cycle.period = 75,
#'                         dates = dates, int.date = int.date,
#'                         s0.r = 0.01*diag(2), s0.eps = 0.1*diag(2),
#'                         niter = 100, burn = 10)
#'
#' # Plotting
#' oldpar <- par(no.readonly = TRUE)
#' par(mfrow=c(2,4))
#' plot(causal.2, type = 'ppchecks', int.date = int.date)
#' par(mfrow=c(2,2))
#' plot(causal.2, type = c('impact','forecast'), int.date = int.date)
#' par(oldpar)


plot.CausalMBSTS <- function(x, int.date, type = c("impact", "forecast", "ppchecks"), prob = NULL, ...) {

    ## Parameter checks
    if(!all(type %in% c("impact", "forecast", "ppchecks", "inclusion.probs")))
        stop("allowed 'type' values are 'impact', 'forecast', 'ppchecks' and 'inclusion.probs'")

    ### Causal effect plot
    if ("impact" %in% type) {
        plotImpact(x, int.date = int.date)
    }

    ### Plot Observed vs Forecast
    if ("forecast" %in% type) {
        plotForecast(x, int.date = int.date)
    }

    ### Posterior predictive checks
    if ("ppchecks" %in% type) {
        plotChecks(x, int.date = int.date)
    }

    # Regressor index plot
    if ("inclusion.probs" %in% type) {
        plotInclusionProb(x, prob = prob)
    }

    return(invisible())
}

# ----------------------------------------------------------------------------------------

#' @import graphics
#'
plotImpact <- function(CausalMBSTS, int.date) {
    dates <- CausalMBSTS$dates
    dim <- dim(CausalMBSTS$mean.general)

    if (!is.null(dim)) {
        d <- dim[2]
        for (i in 1:d) {
            ylim <- c(min(CausalMBSTS$lower.general[, i]), max(CausalMBSTS$upper.general[, i]))
            x <- dates[dates >= int.date]
            main <- paste("Pointwise effect ", "Y", i, sep = "")
            plot(y = CausalMBSTS$upper.general[, i], x = x, type = "l", ylim = ylim,
                 main = main, ylab = "", xlab = "")
            lines(y = CausalMBSTS$lower.general[, i], x = x)
            polygon(c(x, rev(x)), c(CausalMBSTS$lower.general[, i], rev(CausalMBSTS$upper.general[, i])),
                    col = "grey", border = "grey")
            lines(y = CausalMBSTS$mean.general[, i], x = x, col = "blue")

        }
    } else {
        for (j in 1:length(CausalMBSTS$mean.general)) {
            dim <- dim(CausalMBSTS$mean.general[[j]])
            start <- which(dates == int.date)
            end <- start + dim[1]
            x <- dates[start:(end - 1)]
            d <- dim[2]
            for (i in 1:d) {
                ylim <- c(min(CausalMBSTS$lower.general[[j]][, i]), max(CausalMBSTS$upper.general[[j]][,i]))
                main <- paste("Pointwise effect ", "Y", i, ", horizon ", j, sep = "")
                plot(y = CausalMBSTS$upper.general[[j]][, i], x = x, type = "l", ylim = ylim,
                     main = main, ylab = "", xlab = "")
                lines(y = CausalMBSTS$lower.general[[j]][, i], x = x)
                polygon(c(x, rev(x)), c(CausalMBSTS$lower.general[[j]][, i], rev(CausalMBSTS$upper.general[[j]][, i])),
                        col = "grey", border = "grey")
                lines(CausalMBSTS$mean.general[[j]][, i], x = x, col = "blue")

            }
        }
    }
}


#------------------------------------------------------------------------------------------------

#' @import graphics
#'
plotForecast <- function(CausalMBSTS, int.date) {
    dates <- CausalMBSTS$dates
    y <- CausalMBSTS$y
    start <- which(dates == int.date) - round(0.4 * sum(dates < int.date))
    post.mean <- apply(CausalMBSTS$predict$post.pred, c(1, 2), mean)

    if (!is.list(CausalMBSTS$mean.general)) {
        for (i in 1:dim(y)[2]) {
            end <- dim(y)[1]
            x <- dates[start:end]
            yi <- y[start:end, i]
            ylim <- c(min(yi, post.mean[start:end, i]), max(yi, post.mean[start:end, i]))
            main <- paste("Forecasted series ", "Y", i, sep = "")
            plot(y = yi, x = x, type = "l", ylim = ylim, ylab = "", xlab = "", main = main, col = "grey")
            lines(post.mean[start:end, i], col = "blue", x = x)
            abline(v = int.date, col = "red")
        }
    } else {
        for (j in 1:length(CausalMBSTS$mean.general)) {
            dim <- dim(CausalMBSTS$mean.general[[j]])
            end <- which(dates == int.date) + dim[1]
            x <- dates[start:(end - 1)]
            d <- dim[2]
            for (i in 1:d) {
                yi <- y[start:(end - 1), i]
                ylim <- c(min(yi, post.mean[start:(end - 1), i]), max(yi, post.mean[start:(end -
                  1), i]))
                main <- paste("Forecasted series ", "Y", i, ", horizon ", j, sep = "")
                plot(y = yi, x = x, type = "l", ylim = ylim, ylab = "", xlab = "", main = main, col = "grey")
                lines(post.mean[start:(end - 1), i], col = "blue", x = x)
                abline(v = int.date, col = "red")
            }
        }
    }
}

#--------------------------------------------------------------------------------------------

#' @import graphics stats
#'
plotChecks <- function(CausalMBSTS, int.date) {
    dates <- CausalMBSTS$dates
    ind <- dates < int.date
    y <- CausalMBSTS$y
    post.pred <- CausalMBSTS$predict$post.pred.0
    post.pred.mean <- apply(post.pred, c(1, 2), mean)

    for (i in 1:dim(y)[2]) {
        # Density of posterior mean vs density of the data before intervention
        plot(density(y[ind, i]), xlab = "", ylab = "",
             main = paste0("Density comparison, Y", i))
        lines(density(post.pred.mean[, i]), col = "blue")

        # Histograms & Bayesian p-value
        max.distrib <- apply(post.pred, c(2, 3), max)
        pvalue <- sum(max.distrib[i, ] >= max(y[ind, i]))/ncol(max.distrib)
        hist(max.distrib[i, ], 30, col = "lightblue", border = "grey", main = paste0("Bayesian p-value = ",
            round(pvalue, 2), ", Y", i), xlab = "Max. in-sample forecasts")
        abline(v = max(y[ind, i]), col = "darkblue", lwd = 3)

        # Residual plots
        y.rep <- matrix(y[ind, i], nrow(y[ind, ]), (CausalMBSTS$mcmc$niter - CausalMBSTS$mcmc$burn),
            byrow = FALSE)
        res <- (y.rep - (post.pred[, i, ] - CausalMBSTS$mcmc$eps.samples[, i, ]))
        std.res <- t(apply(res, 1, FUN = "/", sqrt(CausalMBSTS$mcmc$Sigma.eps[i, i, ])))
        qqnorm(rowMeans(std.res), main = paste0("Residual QQ-plot, Y", i))
        qqline(rowMeans(std.res))
        Acf(rowMeans(std.res), main = paste0("Residual ACF, Y", i))
    }
}

#-----------------------------------------------------------------------------------------------

#' @import graphics
#'
plotInclusionProb <- function(CausalMBSTS, prob = .5) {
    if (is.null(prob))
        {
            prob <- 0.5
        }  # check here: if no reg above that prob threshold is present stop and write a message (e.g. no reg above that threshold, try to lower the threshold)
    X <- CausalMBSTS$mcmc$X
    select <- apply(CausalMBSTS$mcmc$Z.beta, 2, mean) > prob  # selecting regressors' with prob of inclusion bigger than some amount

    plot(apply(CausalMBSTS$mcmc$Z.beta, 2, mean)[select], type = "h", lwd = 2,
         ylab = expression(paste("Pr(", italic(z[j] == 1), "|", italic(y), ",X)", sep = "")),
         las = 2, xaxt = "none", xlab = "", main = "Regressors' inclusion probabilities")
    axis(1, seq(1, ncol(X[, select]), 1), las = 2, labels = colnames(X)[select])
}
